File: scripts/warm_tile_cache.mjs
#!/usr/bin/env node
/**
 * Pre-warm nginx tile cache for MapServer WMS charts.
 *
 * Uses the EXACT same bbox calculation as Leaflet (V8 JavaScript engine),
 * guaranteeing bit-identical URLs so the nginx fastcgi_cache key matches
 * browser requests perfectly.
 *
 * Tile coverage:
 *   Zoom  1-12 : Wide PNW area  (lat 36–58, lon -138 to -120) — fast, ~80k tiles
 *   Zoom 13-15 : ENC data bounds (lat 45.8–49.1, lon -124.7 to -122.3) — ~150k tiles
 *
 * Usage:
 *   node warm_tile_cache.mjs [--max-zoom 15] [--workers 8]
 *   tail -f /tmp/warm_tiles.log
 */

import http from 'http';
import { parseArgs } from 'util';

// ── WMS parameters (must match sbs-chart.js exactly) ────────────────────────
const LAYERS =
  'DEPARE_verydeep,DEPARE_deep,DEPARE_mid,DEPARE_shallow,DEPARE_vshallow,' +
  'DEPARE_drying,DEPARE_neg,SBDARE,LNDARE,DRGARE,DEPCNT,COALNE,SLCONS,' +
  'WRECKS,OBSTRN,UWTROC,SOUNDG';

const LAYERS_ENC = LAYERS.replace(/,/g, '%2C');

// Host: run locally on Mac → requests go to sailboatserver.local over LAN
// This ensures the bbox computation uses the same x86-64 V8 as Chrome (not ARM64 Pi),
// which is required for the nginx cache key to match real browser requests.
const HOST = process.env.TILE_HOST || 'sailboatserver.local';

const URL_PREFIX =
  `http://${HOST}/cgi-bin/mapserv` +
  '?service=WMS' +
  '&request=GetMap' +
  `&layers=${LAYERS_ENC}` +
  '&styles=' +
  '&format=image%2Fpng' +
  '&transparent=true' +
  '&version=1.1.1' +
  '&width=256' +
  '&height=256' +
  '&srs=EPSG%3A3857' +
  '&bbox=';

// ── Geographic bounds ────────────────────────────────────────────────────────
const ENC_BOUNDS  = { xmin: -124.656917, ymin: 45.861477, xmax: -122.292914, ymax: 49.004167 };
const WIDE_BOUNDS = { xmin: -138.0,      ymin: 36.0,      xmax: -120.0,      ymax: 58.0 };
const WIDE_MAX_ZOOM = 12;

// ── Leaflet EPSG:3857 Transformation (bit-identical to browser) ───────────────
// From Leaflet source: transformation: ht(lt=0.5/(Math.PI*R), 0.5, -lt, 0.5)
// untransform: x = (px/scale - _b) / _a,  y = (py/scale - _d) / _c
const R = 6378137.0;
const _a =  0.5 / (Math.PI * R);
const _c = -0.5 / (Math.PI * R);

function tileBbox(z, x, y) {
  const scale = 256 * Math.pow(2, z);
  const xmin = (x * 256 / scale - 0.5) / _a;
  const xmax = ((x + 1) * 256 / scale - 0.5) / _a;
  const ymax = (y * 256 / scale - 0.5) / _c;       // NW = north = larger y
  const ymin = ((y + 1) * 256 / scale - 0.5) / _c; // SE = south = smaller y
  return [xmin, ymin, xmax, ymax];
}

function makeUrl(z, x, y) {
  const [xmin, ymin, xmax, ymax] = tileBbox(z, x, y);
  return `${URL_PREFIX}${xmin},${ymin},${xmax},${ymax}`;
}

// ── Tile coordinate helpers ───────────────────────────────────────────────────
function lonToX(lon, z) {
  return Math.floor((lon + 180) / 360 * Math.pow(2, z));
}
function latToY(lat, z) {
  const latRad = lat * Math.PI / 180;
  const n = Math.pow(2, z);
  return Math.floor((1 - Math.log(Math.tan(latRad) + 1 / Math.cos(latRad)) / Math.PI) / 2 * n);
}

function* tilesForBounds(bounds, z) {
  const x0 = lonToX(bounds.xmin, z);
  const x1 = lonToX(bounds.xmax, z);
  const y0 = latToY(bounds.ymax, z); // north → smaller y
  const y1 = latToY(bounds.ymin, z); // south → larger y
  for (let x = x0; x <= x1; x++)
    for (let y = y0; y <= y1; y++)
      yield [z, x, y];
}

// ── HTTP request with promise ────────────────────────────────────────────────
function fetchTile(z, x, y) {
  return new Promise((resolve) => {
    const url = makeUrl(z, x, y);
    const req = http.get(url, { headers: { 'User-Agent': 'TileWarmer/1.0' } }, (res) => {
      let size = 0;
      res.on('data', (chunk) => { size += chunk.length; });
      res.on('end', () => {
        const cacheStatus = res.headers['x-cache-status'] || '?';
        resolve({ z, x, y, size, cacheStatus, err: null });
      });
    });
    req.setTimeout(300000);
    req.on('error', (e) => resolve({ z, x, y, size: 0, cacheStatus: 'ERR', err: e.message }));
    req.on('timeout', () => { req.destroy(); resolve({ z, x, y, size: 0, cacheStatus: 'TIMEOUT', err: 'timeout' }); });
  });
}

// ── Self-test: verify bbox matches known tiles from nginx access.log ──────────
function selfTest() {
  const known = [
    [6,  6, 25, [-16280475.528516265, 3757032.8142729844, -15654303.3928041,    4383204.949985149]],
    [6, 12, 22, [-12523442.714243278, 5635549.221409476,  -11897270.578531114,  6261721.357121641]],
    [6, 12, 25, [-12523442.714243278, 3757032.8142729844, -11897270.578531114,  4383204.949985149]],
    [6, 12, 23, [-12523442.714243278, 5009377.085697314,  -11897270.578531114,  5635549.221409476]],
    [6, 12, 20, [-12523442.714243278, 6887893.492833805,  -11897270.578531114,  7514065.628545967]],
  ];
  let allPass = true;
  for (const [z, x, y, expected] of known) {
    const got = tileBbox(z, x, y);
    const match = got.every((v, i) => v === expected[i]);
    if (!match) {
      allPass = false;
      console.log(`MISMATCH z=${z} x=${x} y=${y}`);
      console.log(`  got:    [${got.join(', ')}]`);
      console.log(`  expect: [${expected.join(', ')}]`);
    }
  }
  return allPass;
}

// ── Worker pool (limit concurrent requests to maxWorkers) ────────────────────
async function runWithConcurrency(tiles, maxWorkers, onResult) {
  let idx = 0;
  async function worker() {
    while (idx < tiles.length) {
      const tile = tiles[idx++];
      const result = await fetchTile(...tile);
      onResult(result);
    }
  }
  await Promise.all(Array.from({ length: maxWorkers }, worker));
}

// ── Main ─────────────────────────────────────────────────────────────────────
async function main() {
  const { values } = parseArgs({
    options: {
      'max-zoom': { type: 'string', default: '15' },
      'workers':  { type: 'string', default: '8' },
      'test':     { type: 'boolean', default: false },
    }
  });

  if (values.test) {
    console.log('Running self-test against nginx access.log tiles...');
    const pass = selfTest();
    console.log(pass ? '✓ All tiles match exactly!' : '✗ Some tiles differ!');
    return;
  }

  const maxZoom   = parseInt(values['max-zoom']);
  const maxWorkers = parseInt(values.workers);

  // Collect all tiles
  const tiles = [];
  const zoomCounts = {};
  for (let z = 1; z <= maxZoom; z++) {
    const bounds = z <= WIDE_MAX_ZOOM ? WIDE_BOUNDS : ENC_BOUNDS;
    let count = 0;
    for (const tile of tilesForBounds(bounds, z)) { tiles.push(tile); count++; }
    zoomCounts[z] = count;
  }

  const total = tiles.length;
  console.log(`\nPre-warming ${total.toLocaleString()} tiles (zoom 1-${maxZoom}, ${maxWorkers} workers)`);
  console.log(`Wide area (zoom 1-${WIDE_MAX_ZOOM}): lon ${WIDE_BOUNDS.xmin}..${WIDE_BOUNDS.xmax}, lat ${WIDE_BOUNDS.ymin}..${WIDE_BOUNDS.ymax}`);
  console.log(`ENC area  (zoom ${WIDE_MAX_ZOOM+1}-${maxZoom}): lon ${ENC_BOUNDS.xmin}..${ENC_BOUNDS.xmax}, lat ${ENC_BOUNDS.ymin}..${ENC_BOUNDS.ymax}`);
  console.log('Zoom breakdown:');
  for (let z = 1; z <= maxZoom; z++) {
    const label = z <= WIDE_MAX_ZOOM ? 'WIDE' : 'ENC ';
    console.log(`  z${String(z).padStart(2)} (${label}): ${String(zoomCounts[z]).padStart(7)} tiles`);
  }
  const estMin = total / maxWorkers / 10;
  if (total > 50000) console.log(`\nEstimated time: ~${(estMin/60).toFixed(0)} hours`);
  console.log(`\nLogging progress every 1000 tiles. Starting...`);

  let done = 0, hit = 0, miss = 0, errors = 0;
  const startMs = Date.now();
  const logLines = [];

  function onResult(r) {
    done++;
    if (r.err) errors++;
    else if (r.cacheStatus.includes('HIT')) hit++;
    else miss++;

    if (done % 1000 === 0 || done === total) {
      const elapsed = (Date.now() - startMs) / 1000;
      const rate = done / elapsed;
      const eta = (total - done) / rate;
      const msg = `  ${done.toLocaleString()}/${total.toLocaleString()} ` +
        `(${(100*done/total).toFixed(1)}%) | ` +
        `HIT=${hit.toLocaleString()} MISS=${miss.toLocaleString()} ERR=${errors} | ` +
        `${rate.toFixed(1)} tiles/s | ETA ${(eta/60).toFixed(0)}m`;
      console.log(msg);
    }
  }

  await runWithConcurrency(tiles, maxWorkers, onResult);

  const elapsed = (Date.now() - startMs) / 1000;
  const summary = `\nDone! ${total.toLocaleString()} tiles in ${(elapsed/60).toFixed(1)} minutes.\n` +
    `Cache hits: ${hit.toLocaleString()} | New tiles: ${miss.toLocaleString()} | Errors: ${errors}`;
  console.log(summary);
}

main().catch(console.error);

File: scripts/warm_tile_cache.py
#!/usr/bin/env python3
"""
Pre-warm nginx tile cache for MapServer WMS charts.

Generates tile requests that EXACTLY match Leaflet's URL format so that
browser tile requests hit the nginx fastcgi_cache instead of re-rendering.

Tile coverage:
  - Zoom 1-12 : Wide PNW area  (lat 36-58, lon -138 to -120) — fast, ~80k tiles
  - Zoom 13-15: ENC data bounds (lat 45.8-49.1, lon -124.7 to -122.3) — ~150k tiles

Run: python3 warm_tile_cache.py [--max-zoom 15] [--workers 4] [--log /tmp/warm.log]
Check: tail -f /tmp/warm_tiles.log
"""

import math
import sys
import time
import argparse
import concurrent.futures
import urllib.request
import urllib.error

# ── WMS constants (must match sbs-chart.js exactly) ──────────────────────────
LAYERS = (
    "DEPARE_verydeep,DEPARE_deep,DEPARE_mid,DEPARE_shallow,DEPARE_vshallow,"
    "DEPARE_drying,DEPARE_neg,SBDARE,LNDARE,DRGARE,DEPCNT,COALNE,SLCONS,"
    "WRECKS,OBSTRN,UWTROC,SOUNDG"
)
# URL-encoded layers string (Leaflet uses encodeURIComponent on the value)
LAYERS_ENC = LAYERS.replace(",", "%2C")

# Static URL prefix — matches Leaflet's getParamString output exactly
# Order: service, request, layers, styles, format, transparent, version, width, height, srs
# (width/height set in initialize; srs set in onAdd — as seen in Leaflet source)
URL_PREFIX = (
    "http://localhost/cgi-bin/mapserv"
    "?service=WMS"
    "&request=GetMap"
    f"&layers={LAYERS_ENC}"
    "&styles="
    "&format=image%2Fpng"
    "&transparent=true"
    "&version=1.1.1"
    "&width=256"
    "&height=256"
    "&srs=EPSG%3A3857"
    "&bbox="
)

# ── Geographic bounds ─────────────────────────────────────────────────────────
# ENC data coverage (from enc_merged.gpkg)
ENC_BOUNDS = (-124.656917, 45.861477, -122.292914, 49.004167)   # xmin,ymin,xmax,ymax

# Wide area — ~500nm from Seattle (covers all likely PNW sailing waters)
WIDE_BOUNDS = (-138.0, 36.0, -120.0, 58.0)

# ── Earth constants (must match Leaflet's R = 6378137 exactly) ───────────────
R = 6378137.0

# Leaflet EPSG:3857 Transformation coefficients:
#   transformation: ht(_a = 0.5/(PI*R), _b = 0.5, _c = -0.5/(PI*R), _d = 0.5)
# untransform: x = (px/scale - _b) / _a,  y = (py/scale - _d) / _c
_LEAFLET_A =  0.5 / (math.pi * R)   # positive
_LEAFLET_C = -0.5 / (math.pi * R)   # negative


def tile_bbox(z: int, x: int, y: int) -> tuple:
    """
    Compute EPSG:3857 bbox using Leaflet's EXACT code path:
      untransform(pt, scale) → (pt.x/scale - _b) / _a,  (pt.y/scale - _d) / _c
    where _a=0.5/(π*R), _b=0.5, _c=-0.5/(π*R), _d=0.5, scale=256*2^z.

    Using division (not multiplication) matches Leaflet bit-for-bit,
    which is required for nginx fastcgi_cache_key ($request_uri) hits.
    The subsequent unproject→project round-trip is an identity, so we skip it.
    """
    scale = 256.0 * (2 ** z)
    nw_px = x       * 256.0
    se_px = (x + 1) * 256.0
    nw_py = y       * 256.0
    se_py = (y + 1) * 256.0

    xmin = (nw_px / scale - 0.5) / _LEAFLET_A   # west edge (NW x)
    xmax = (se_px / scale - 0.5) / _LEAFLET_A   # east edge (SE x)
    ymax = (nw_py / scale - 0.5) / _LEAFLET_C   # north edge (NW y) — _c is negative
    ymin = (se_py / scale - 0.5) / _LEAFLET_C   # south edge (SE y)
    return xmin, ymin, xmax, ymax


def lon_to_x(lon: float, z: int) -> int:
    return int((lon + 180.0) / 360.0 * (2 ** z))


def lat_to_y(lat: float, z: int) -> int:
    lat_r = math.radians(lat)
    n = 2 ** z
    return int((1.0 - math.log(math.tan(lat_r) + 1.0 / math.cos(lat_r)) / math.pi) / 2.0 * n)


def tiles_for_bounds(bounds: tuple, z: int):
    """Yield all (z, x, y) tile coords covering the given WGS84 bounds."""
    lon_min, lat_min, lon_max, lat_max = bounds
    x0 = lon_to_x(lon_min, z)
    x1 = lon_to_x(lon_max, z)
    y0 = lat_to_y(lat_max, z)  # north → smaller y
    y1 = lat_to_y(lat_min, z)  # south → larger y
    for x in range(x0, x1 + 1):
        for y in range(y0, y1 + 1):
            yield z, x, y


def make_url(z: int, x: int, y: int) -> str:
    xmin, ymin, xmax, ymax = tile_bbox(z, x, y)
    return f"{URL_PREFIX}{xmin},{ymin},{xmax},{ymax}"


def fetch_tile(args):
    z, x, y = args
    url = make_url(z, x, y)
    try:
        req = urllib.request.Request(url, headers={"User-Agent": "TileWarmer/1.0"})
        with urllib.request.urlopen(req, timeout=300) as resp:
            data = resp.read()
            status = resp.getheader("X-Cache-Status", "?")
            return z, x, y, len(data), status, None
    except Exception as e:
        return z, x, y, 0, "ERR", str(e)


def count_tiles(zoom_levels, wide_zooms, enc_zooms):
    total = 0
    for z in zoom_levels:
        bounds = WIDE_BOUNDS if z <= wide_zooms else ENC_BOUNDS
        count = sum(1 for _ in tiles_for_bounds(bounds, z))
        total += count
    return total


def main():
    parser = argparse.ArgumentParser(description="Pre-warm MapServer WMS tile cache")
    parser.add_argument("--max-zoom", type=int, default=15,
                        help="Maximum zoom level to pre-warm (default: 15)")
    parser.add_argument("--workers", type=int, default=4,
                        help="Parallel workers (default: 4, matches fcgiwrap)")
    parser.add_argument("--log", default="/tmp/warm_tiles.log",
                        help="Log file path (default: /tmp/warm_tiles.log)")
    parser.add_argument("--test", action="store_true",
                        help="Test URL format against known tile from nginx log and exit")
    args = parser.parse_args()

    if args.test:
        # Verify our bbox formula against the known tile from nginx access.log:
        # Tile z=6, x=12 has xmin=-12523442.714243278 (verified against Leaflet)
        xmin, ymin, xmax, ymax = tile_bbox(6, 12, 20)
        print(f"Test tile z=6 x=12 y=20:")
        print(f"  xmin = {xmin}")
        print(f"  ymin = {ymin}")
        print(f"  xmax = {xmax}")
        print(f"  ymax = {ymax}")
        print(f"Expected xmin: -12523442.714243278")
        print(f"Expected ymax:  7514065.628545967")
        print(f"Match xmin: {abs(xmin - (-12523442.714243278)) < 1e-3}")
        print(f"Match ymax: {abs(ymax - 7514065.628545967) < 1e-3}")
        print()
        print(f"Sample URL:\n  {make_url(6, 12, 20)}")
        return

    # Split: wide area for zoom ≤ 12, ENC bounds for zoom 13+
    WIDE_MAX_ZOOM = 12
    zoom_levels = list(range(1, args.max_zoom + 1))

    # Pre-count tiles
    print("Counting tiles...", flush=True)
    tile_list = []
    zoom_counts = {}
    for z in zoom_levels:
        bounds = WIDE_BOUNDS if z <= WIDE_MAX_ZOOM else ENC_BOUNDS
        ztiles = list(tiles_for_bounds(bounds, z))
        zoom_counts[z] = len(ztiles)
        tile_list.extend(ztiles)

    total = len(tile_list)
    print(f"\nPre-warming {total:,} tiles (zoom 1-{args.max_zoom}, {args.workers} workers)")
    print(f"Wide area coverage (zoom 1-{WIDE_MAX_ZOOM}): {WIDE_BOUNDS}")
    print(f"ENC data coverage  (zoom {WIDE_MAX_ZOOM+1}-{args.max_zoom}): {ENC_BOUNDS}")
    print(f"Zoom breakdown:")
    for z in zoom_levels:
        bounds_label = "WIDE" if z <= WIDE_MAX_ZOOM else "ENC "
        print(f"  z{z:2d} ({bounds_label}): {zoom_counts[z]:6,} tiles")
    if total > 50000:
        est_min = total / args.workers / 10  # ~10 tiles/sec per worker pessimistic
        print(f"\nEstimated time: ~{est_min/60:.0f} hours (runs in background, check {args.log})")
    print(f"\nLogging to: {args.log}")
    print("Starting... (Ctrl-C to stop, already-cached tiles are instant)\n", flush=True)

    done = 0
    hit = 0
    miss = 0
    errors = 0
    start = time.time()

    with open(args.log, "w") as logf:
        logf.write(f"TileWarmer started: {total} tiles, zoom 1-{args.max_zoom}\n")
        logf.flush()

        with concurrent.futures.ThreadPoolExecutor(max_workers=args.workers) as executor:
            futures = {executor.submit(fetch_tile, t): t for t in tile_list}
            for future in concurrent.futures.as_completed(futures):
                z, x, y, size, status, err = future.result()
                done += 1
                if err:
                    errors += 1
                elif "HIT" in status:
                    hit += 1
                else:
                    miss += 1

                if done % 500 == 0 or done == total:
                    elapsed = time.time() - start
                    rate = done / elapsed if elapsed > 0 else 0
                    eta_s = (total - done) / rate if rate > 0 else 0
                    msg = (
                        f"  {done:7,}/{total:,} ({100*done/total:5.1f}%) "
                        f"| HIT={hit:,} MISS={miss:,} ERR={errors} "
                        f"| {rate:5.1f} tiles/s "
                        f"| ETA {eta_s/60:.0f}m\n"
                    )
                    sys.stdout.write(msg)
                    sys.stdout.flush()
                    logf.write(msg)
                    logf.flush()

                if err and errors <= 10:
                    logf.write(f"    ERR z={z} x={x} y={y}: {err}\n")
                    logf.flush()

    elapsed = time.time() - start
    summary = (
        f"\nDone! {total:,} tiles in {elapsed/60:.1f} minutes.\n"
        f"Cache hits: {hit:,} | New tiles rendered: {miss:,} | Errors: {errors}\n"
    )
    print(summary)
    with open(args.log, "a") as logf:
        logf.write(summary)


if __name__ == "__main__":
    main()
